tsdb/wal/wal.go

```diff
@@ -757,23 +769,52 @@ func listSegments(dir string) (refs []segmentRef, err error) {
 		return nil, err
 	}
 	var last int
+	// assume segments in sequence
+	inSequence := true
 	for _, fn := range files {
 		k, err := strconv.Atoi(fn)
 		if err != nil {
 			continue
 		}
-		if len(refs) > 0 && k > last+1 {
-			return nil, errors.New("segments are not sequential")
+
+		if len(refs) > 0 && k > last+1 && inSequence {
+			inSequence = false
 		}
 		refs = append(refs, segmentRef{name: fn, index: k})
 		last = k
 	}
+	if !inSequence {
+		return refs, errors.New(ErrorUnsequentialSegments)
+	}
 	sort.Slice(refs, func(i, j int) bool {
 		return refs[i].index < refs[j].index
 	})
 	return refs, nil
 }
 
+// Returns the sequential segments after deleting the out of sequence wals after the sequence gap.
+func (w *WAL) repairUnsequentialSegments() error {
+	segments, _ := listSegments(w.dir)
+	level.Warn(w.logger).Log("msg", "deleting unsequential segments")
+	last, outofSequence := -1, false
+
+	for _, seg := range segments {
+		k, err := strconv.Atoi(seg.name)
+		if err != nil {
+			return errors.Wrap(err, "faulty segment name")
+		}
+		if k > last+1 || outofSequence {
```

Comment:
When the segments are ordered by name you can just simply delete all the segments after the last good segment.
Something like:

```
if k > last+1 {
	log which is the last good segment and the segments which will be deleted.
	_,seg:=range segments[k:]{
		err := os.Remove(SegmentName(w.dir, seg.index))
		if err != nil {
			return errors.Wrapf(err, "delete segment: %s", seg.name)
		}
	}
	return nil	
}

```